### PR TITLE
Document Azure CLI access

### LINKED ISSUE #17583: Document Azure CLI access via Teleport
## Applies To
This will likely be a guide in `/docs/application-access`.

## Details
When we add Azure Console support for Application Access (as tracked in https://github.com/gravitational/teleport/issues/11780), we'll need to document it.

## Related Issues

- https://github.com/gravitational/teleport/issues/11780
- https://github.com/gravitational/teleport/issues/16401

### PR DESCRIPTION
**Pull request by ptgott**
Closes #17583

- Create a guide to Azure CLI access
- Update references to reflect changes to roles, commands, and the file config

**TODO** Remember to backport to v12 when this is available

### CODE REVIEW

**Comment by ptgott** [comment]
@Tener From what I gather, the `tsh proxy azure` command requires a registered Azure app and is intended for custom Azure apps (and not, say, `curl`). I didn't include it in the step-by-step guide because it would require a custom Azure app. Let me know if there would be a simple way to add this command to the step-by-step guide. Either way, I've described this command in the CLI reference.

File: docs/pages/application-access/controls.mdx
Comment (alexfornuto) [inline docs/pages/application-access/controls.mdx:122]: In the command this is referenced as `MY_RESOURCE_GROUP`, not sure which way you want to change it.

File: docs/pages/application-access/guides/azure.mdx
Comment (alexfornuto) [inline docs/pages/application-access/guides/azure.mdx:16]: Is this instead of the declaration above "The Teleport Application Service runs on an Azure VM...", or does it refer to an Azure VM in a private network (not publicly accessible), as opposed to a VM in a physical private network like an in-office LAN?

File: docs/pages/application-access/guides/azure.mdx
Comment (alexfornuto) [inline docs/pages/application-access/guides/azure.mdx:170]: The tenses conflict here as "are running" compared to "will install" below.

File: docs/pages/application-access/guides/azure.mdx
Comment (alexfornuto) [inline docs/pages/application-access/guides/azure.mdx:284]: Could/should we specify?

```suggestion
copied in [step 1](#create-an-azure-managed-identity).
```

File: docs/pages/application-access/guides/azure.mdx
Comment (alexfornuto) [inline docs/pages/application-access/guides/azure.mdx:292]: Assuming your [other PR](https://github.com/gravitational/teleport/pull/18831) is merged by the time you come back to this, I suggest replacing this with your cool new partial.

File: docs/pages/application-access/guides/azure.mdx
Comment (alexfornuto) [inline docs/pages/application-access/guides/azure.mdx:455]: This seems technically correct and all that, but it's a mouthful and took me a couple passes to parse. Not sure how to best rephrase, maybe...

```suggestion
You must attach all identities you would like users to access to the Azure VM
running the Application Service. This lets the Application Service make requests to
the Azure API using those identities.
```

Something like that, splitting up the single sentence?

File: docs/pages/application-access/guides/azure.mdx
Comment (alexfornuto) [inline docs/pages/application-access/guides/azure.mdx:493]: Would they live it blank or not include the flag?

**Review by alexfornuto — APPROVED** [review]
Reviewed for copy only. Comments and suggestions below.

Additionally, what do you think about a tip for users that will exclusively use the `az` tool through `tsh` to create an alias in their shell rc file?

File: docs/pages/application-access/guides/azure.mdx
Comment (Tener) [reply docs/pages/application-access/guides/azure.mdx:493]: If they have access to exactly one identity, the flag can be omitted. Otherwise, you need to specify the identity. Perhaps this paragraph could be rephrased to educate about the reason the flag is sometimes required and sometimes not.

**Comment by Tener** [comment]
> @Tener From what I gather, the `tsh proxy azure` command requires a registered Azure app and is intended for custom Azure apps (and not, say, `curl`). I didn't include it in the step-by-step guide because it would require a custom Azure app. Let me know if there would be a simple way to add this command to the step-by-step guide. Either way, I've described this command in the CLI reference.

Hmm, not really, `tsh proxy azure` is usable whenever `tsh az` is. In general terms, you can say that `tsh proxy azure` is more flexible but requires a notch more setup.

In particular, you must manage the lifetime of the proxy and manually set the env variables shown:

```
$ tsh proxy azure --port 3500
Started Azure proxy on http://127.0.0.1:3500.

Use the following credentials and HTTPS proxy setting to connect to the proxy:

  export AZURE_CONFIG_DIR=/Users/tener/.azure-teleport-azure
  export HTTPS_PROXY=http://127.0.0.1:3500
  export MSI_ENDPOINT=https://azure-msi.teleport.dev
  export REQUESTS_CA_BUNDLE=/Users/tener/.tsh/keys/boson.tener.io/tener-app/boson.tener.io/azure-localca.pem
```

Now in another console:
```
$ export AZURE_CONFIG_DIR=/Users/tener/.azure-teleport-azure
$ export HTTPS_PROXY=http://127.0.0.1:3500
$ export MSI_ENDPOINT=https://azure-msi.teleport.dev
$ export REQUESTS_CA_BUNDLE=/Users/tener/.tsh/keys/boson.tener.io/tener-app/boson.tener.io/azure-localca.pem

$ az vm list | jq '.[].vmId'
"fecd46c9-202c-4325-b3f2-196a73e5eb3f"
"d52e08a0-dc5c-44b3-8832-5fd33105f342"

$ az account show | jq '.user'
{
  "assignedIdentityInfo": "MSIResource-/subscriptions/060a97ea-3a57-4218-9be5-dba3f19ff2b5/resourceGroups/tener/providers/Microsoft.ManagedIdentity/userAssignedIdentities/tener-test-2",
  "name": "userAssignedIdentity",
  "type": "servicePrincipal"
}
```

If all you want to use is `az`, then `tsh proxy azure` is probably not worth the hassle; `tsh az` should work just fine. But the important point is that once the environment variables are set, you don't need any wrapper around `az` command. This is very useful for scripts that would otherwise need to be changed to accommodate a different command being used.  Another important use case is _other_ applications making use of Azure SDK - they should detect the credentials present in env variables and make appropriate use of those.

File: docs/pages/application-access/controls.mdx
Comment (Tener) [reply docs/pages/application-access/controls.mdx:91]: Taking this a step further, we can advise a command to list available identities. If you have an Azure VM with assigned identities you can run:
```
# login with vm-assigned managed identity
$ az login --identity
# list available user assigned identities
$ az identity list
```
OTOH not everyone will want to install `az` on their machines, so perhaps this isn't the best way to list these identities.

File: docs/pages/application-access/guides/azure.mdx
Comment (Tener) [inline docs/pages/application-access/guides/azure.mdx:519]: This is now swapped; first we print the output from `az login -i`, then the comments from `tsh`.

```suggestion
[
  {
    "environmentName": "AzureCloud",
    "homeTenantId": "00000000-0000-0000-0000-000000000000",
    "id": "00000000-0000-0000-0000-000000000000",
    "isDefault": true,
    "managedByTenants": [],
    "name": "Microsoft Azure Sponsorship",
    "state": "Enabled",
    "tenantId": "00000000-0000-0000-0000-000000000000",
    "user": {
      "assignedIdentityInfo": "MSIResource-/subscriptions/0000000000000-0000-0000-000000000000/resourceGroups/example-resource-group/providers/Microsoft.ManagedIdentity/userAssignedIdentities/teleport-azure",
      "name": "userAssignedIdentity",
      "type": "servicePrincipal"
    }
  }
]

Logged into Azure app "azure-cli".
Your identity: /subscriptions/0000000000000-0000-0000-000000000000/resourceGroups/example-resource-group/providers/Microsoft.ManagedIdentity/userAssignedIdentities/teleport-azure
Example Azure CLI command: tsh az vm list
```

File: docs/pages/application-access/guides/azure.mdx
Comment (Tener) [inline docs/pages/application-access/guides/azure.mdx:529]:
```suggestion
$ tsh az vm list
```

File: docs/pages/application-access/reference.mdx
Comment (Tener) [inline docs/pages/application-access/reference.mdx:32]:
```suggestion
    # Optional. For access to cloud provider APIs, specify the cloud provider. Allowed
```

File: docs/pages/reference/config.mdx
Comment (Tener) [inline docs/pages/reference/config.mdx:756]:
```suggestion
      # Optional. For access to cloud provider APIs, specify the cloud provider. Allowed
```

File: docs/pages/reference/cli.mdx
Comment (Tener) [inline docs/pages/reference/cli.mdx:1161]: We are missing `--azure-identities` flag for `tctl users add`.

File: docs/pages/application-access/controls.mdx
Comment (ptgott) [reply docs/pages/application-access/controls.mdx:91]: Thanks! Would it be more straightforward to recommend visiting the "Identity" tab in the page for the VM in the Azure Portal so users won't need to install `az` on the VM?

File: docs/pages/application-access/guides.mdx
Comment (ptgott) [reply docs/pages/application-access/guides.mdx:9]: 👍 I think I'd neglected to delete that "provide" while editing—it shouldn't be there.

File: docs/pages/application-access/guides/azure.mdx
Comment (ptgott) [reply docs/pages/application-access/guides/azure.mdx:16]: I've added some context—this refers to running the VM without exposing any ports to the public internet

File: docs/pages/application-access/guides/azure.mdx
Comment (ptgott) [reply docs/pages/application-access/guides/azure.mdx:284]: I've specified the step, though I don't think we should create a link, since that could lead to the reader landing in some possibly unexpected part of the doc and losing their place.

File: docs/pages/application-access/guides/azure.mdx
Comment (ptgott) [reply docs/pages/application-access/guides/azure.mdx:493]: 👍

File: docs/pages/application-access/guides/azure.mdx
Comment (ptgott) [reply docs/pages/application-access/guides/azure.mdx:14]: I've added more context

File: docs/pages/application-access/guides/azure.mdx
Comment (ptgott) [reply docs/pages/application-access/guides/azure.mdx:141]: I've added these instructions using the Azure Portal since the previous instructions also used this

File: docs/pages/application-access/guides/azure.mdx
Comment (ptgott) [reply docs/pages/application-access/guides/azure.mdx:146]: Since this section is about running the Application Service, and I've added more context about how the Application Service works at the beginning of the guide, I've suggested removing everything in this paragraph after the first sentence.

File: docs/pages/application-access/guides/azure.mdx
Comment (ptgott) [reply docs/pages/application-access/guides/azure.mdx:150]: Our step-by-step guides for Teleport's services usually take the manual approach (I think the only exception is the Desktop Access [getting started guide](https://goteleport.com/docs/desktop-access/getting-started/)). I definitely think we need to include instructions for using the automatic approach, but am not sure what the cleanest way to do so is. We have a [GitHub issue for this](https://github.com/gravitational/teleport/issues/15060), and I am tempted to leave this task for when we start working on that issue.

**Comment by ptgott** [comment]
@alexfornuto I'm making pretty substantial edits while responding to feedback, so you may want to wait until @Tener signs off on this before giving it a second look. We'll probably also have to optimize images again, since I'm changing/adding screenshots.

File: docs/pages/application-access/guides/azure.mdx
Comment (ptgott) [reply docs/pages/application-access/guides/azure.mdx]: Since we're using the new `add-role-to-user.mdx` partial, I've opened an issue to use `tctl users update` in that partial: https://github.com/gravitational/teleport/issues/19067

File: docs/pages/application-access/guides/azure.mdx
Comment (ptgott) [reply docs/pages/application-access/guides/azure.mdx:302]: I'm not super familiar with SSO in Teleport—would this mean defining a custom claim/attribute via your IdP called `azure_identities`, then adding `{{external.azure_identities}}` under `allow.azure_identities` in the `azure-cli-access` role?

File: docs/pages/application-access/guides/azure.mdx
Comment (Tener) [inline docs/pages/application-access/guides/azure.mdx:452]: This is being discussed in the code PR, but most likely, we will have some auth mechanism for accessing `MSI_ENDPOINT`: https://github.com/gravitational/teleport/pull/18675#discussion_r1040641188. 

The likely solution is a random hex attached at the end:
```
export MSI_ENDPOINT=https://azure-msi.teleport.dev/eedfd5b55257c0aaa58f
```

But also with an option to set this by the user:
```
$ export MSI_ENDPOINT=https://azure-msi.teleport.dev/MY-CUSTOM-SECRET
$ tsh proxy azure --port 3500
Started Azure proxy on http://127.0.0.1:3500.

Use the following credentials and HTTPS proxy setting to connect to the proxy:

  export AZURE_CONFIG_DIR=/Users/tener/.azure-teleport-azure
  export HTTPS_PROXY=http://127.0.0.1:3500
  export MSI_ENDPOINT=https://azure-msi.teleport.dev/MY-CUSTOM-SECRET
  export REQUESTS_CA_BUNDLE=/Users/tener/.tsh/keys/boson.tener.io/tener-app/boson.tener.io/azure-localca.pem
```

This may be one of the details that will have to wait until the code PR is merged to master, though.

**Review by Tener — COMMENTED** [review]
Looks good. The details are still in flux, however, since I'm still receiving feedback for the code PR. Although I doubt the final delta will be large.

File: docs/pages/application-access/controls.mdx
Comment (alexfornuto) [reply docs/pages/application-access/controls.mdx]: I'm happy with the changes; feel free to mark this resolved once the rest of the discussion is finished.

File: docs/pages/application-access/reference.mdx
Comment (zmb3) [inline docs/pages/application-access/reference.mdx:34]: I assume you would not typically set this for an app such as Grafana.

Should we expand the example to add separate `apps` entries for the cloud consoles with more realistic examples?

File: docs/pages/reference/cli.mdx
Comment (Tener) [inline docs/pages/reference/cli.mdx:447]: `AZURE_CONFIG_DIR` is now `~/.tsh/azure/<cluster>/<app-name>`, so something like:

```suggestion
  export AZURE_CONFIG_DIR=/Users/myuser/.tsh/azure/my.teleport.cluster/azure
```

**Review by Tener — APPROVED** [review]
Thanks for the proactive update of the PR @ptgott ! There are only two minor things I've noticed, looks good otherwise.

**Comment by ptgott** [comment]
@alexfornuto Letting you know that this is ready for another look if you'd like to check on anything before we merge. Would you mind running the optimization command again for images? Thanks!

File: docs/pages/application-access/guides/azure.mdx
Comment (alexfornuto) [inline docs/pages/application-access/guides/azure.mdx:196]: Kubernetes service isn't correct for this context, though I'm also not seeing this var being used in the render.

File: docs/pages/application-access/guides/azure.mdx
Comment (alexfornuto) [inline docs/pages/application-access/guides/azure.mdx:212]: Following other discussions in docs PRs and the standard implementation we're suggesting, we should probably assume/define the proxy server with a default port of `443` in our examples. This comment applies to L226 as well.

**Review by alexfornuto — APPROVED** [review]
Just a few nits, 👍 once addressed.

File: docs/pages/application-access/guides/azure.mdx
Comment (ptgott) [reply docs/pages/application-access/guides/azure.mdx:16]: There are, including custom ones. I've clarified what an "Azure CLI" means in the first paragraph.

File: docs/pages/application-access/guides/azure.mdx
Comment (ptgott) [reply docs/pages/application-access/guides/azure.mdx:196]: That was a lazy copy/paste oversight on my part. That partial had a parameter at one point, but I removed it because the docs linter was (at the time, but not now) erroring on partials with default param values.

**Pull request merged**
